### ISSUE TITLE
Change Request: Fix safety with "safe" and "unsafe" fixes

### ISSUE THREAD

**Issue by thernstig** [eslint/eslint#20100]
### ESLint version

v9.35.0

### What problem do you want to solve?

ESLint could benefit from a mechanism to flag certain rule fixes as "unsafe," allowing them to be applied automatically while signaling that manual review is necessary. The inspiration came from [Ruff](https://docs.astral.sh/ruff/), a modern Python linter.

Taken from https://docs.astral.sh/ruff/linter/#fix-safety:

> Ruff labels fixes as "safe" and "unsafe". The meaning and intent of your code will be retained when applying safe fixes, but the meaning could change when applying unsafe fixes.
>
> Specifically, an unsafe fix could lead to a change in runtime behavior, the removal of comments, or both, while safe fixes are intended to preserve runtime behavior and will only remove comments when deleting entire statements or expressions (e.g., removing unused imports).

Here is a trivial, nonsensical example using the rule [`no-unused-vars`](https://eslint.org/docs/latest/rules/no-unused-vars):

```js
function sum(a, b) {
  return a + 1;
}

// Client expects sum to add both numbers
const total = sum(2, 3);  // Expected: 5
console.log(total);       // Actual: 3 (because b is ignored)
```

The author might have forgotten to remove the `b` (and rename the function). ESLint could remove the `b` for us if an "unsafe fix" was applied.

### What do you think is the correct solution?

Add a new `meta.unsafeFixable` to [rules](https://eslint.org/docs/latest/extend/custom-rules). It would be used as:

```sh
# Show unsafe fixes
npx eslint --unsafe-fixes . 

# Apply unsafe fixes
npx eslint --fix --unsafe-fixes . 
```

This enhancement would combine the efficiency of automated fixes with the caution required for complex code modifications.

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

Related concept: https://github.com/eslint/eslint/issues/18696.

**Comment by nzakas** [comment]
Thanks for the suggestion. I don't think this is a good idea. I understand why you might want it, but the trouble comes when an unsafe fix is applied and you have no idea what happened. The amount of time it takes to debug such an issue is non-trivial, and I'm speaking from experience because the earliest version of the `eqeqeq` rule made an unsafe fix that cost me a day of work figuring out what happened.

If anyone else on the team feels differently, we can consider it. As for me, I'm :-1:.

cc @eslint/eslint-team

**Comment by DMartens** [comment]
First for clarification of the example of `no-unused-vars` to make sure we are on the same page:
This does not show an "unsafe" fix as this does not change runtime behavior.
It is a suggestion, as we do not know whether the code is "finished" and the rule is run in an editor on save.
A better example would be the suggestion for `eqeqeq` as changing `==` to `===` can change runtime behavior.

Next for the proposal to use `meta.unsafeFixes`:
Ideally marking a fix as unsafe should be dynamic.
As an example a fictive rule which prefers `Array#includes` over `Array#indexOf`.
The problem is that the object must not be an array to have the `indexOf` method but if the rule can statically determine the value is an array, the fix is safe:
```javascript
const knownArray = [];
knownArray.indexOf(value) !== -1;
```

but if it is not possible (and not types from TypeScript parser services are available) this would be an unsafe fix:
```javascript
function fn(unknown) {
    // unknown could also be a custom object with an `indexOf` method without an `includes` method
	return unknown.indexOf(value) !== -1;
}
```

ESLint uses suggestions for "unsafe" fixes.
So we could implement the dynamic behavior by extending suggestions of `suggest` of `context.report` by adding a `unsafe: boolean` property and support automatically applying suggestions from the CLI.
This would also benefit the non-unsafe case to show that the code is "finished" (e.g. `no-unused-vars` and `no-empty`).

**Comment by thernstig** [comment]
That bot sure is an aggressive closer 😆 Seems it was not wanted anyway.
